### ISSUE TITLE
Provide consistent Precondition behavior for arrays and collections

### ISSUE THREAD

**Issue by britter** [junit-team/junit5#496]
After reading @sbrannen comment in #493 I had a look at the `Preconditions` class. Currently the implementation of `notNull(Object[])` checks that the array reference is not null and that the array does not contain any null values. The counterpart for Collections is `notEmpty(Collection)`, which checks that the given collection is not null and that it is not empty.

Looking at for example the`TagsFilter` class and how it checks its input parameters:

```java
public static PostDiscoveryFilter excludeTags(String... tags) {
	Preconditions.notNull(tags, "tags must not be null");
	return excludeTags(asList(tags));
}

public static PostDiscoveryFilter excludeTags(List<String> tags) {
	Preconditions.notEmpty(tags, "tags must not be null or empty");
	return descriptor -> FilterResult.includedIf(trimmedTagsOf(descriptor).noneMatch(tags::contains));
}
```

We can conclude that:

Method call | Result
--------------------|------------
 `excludeTags(null)` | Exception 
 `excludeTags((String[]) null)` | Exception 
 `excludeTags(new String[0]())` | okay 
 `excludeTags(Collections.singletonList(null))` | okay 
 `excludeTags((List<String>) null)` | Exception 
 `excludeTags(Collections.emptyList())` | Exceptions

My proposal here would be to introduce new methods:

* `Preconditions.noneNull(Collection)`
* `Preconditions.noneNull(Object[])`

These would check that the reference is not null, that the collection/array does not contain any null reference and that the collection/array is not empty. The last check does not really match the method name but I can't think of a better name right now...

**Comment by britter** [comment]
Actually `excludeTags(String[])` with an empty array will throw an exception because it delegates to `excludeTags(Collection)`.

**Comment by sbrannen** [comment]
Thanks for investigating and reporting back on your findings.

I'm slating this for M3 for the time being.

**Comment by britter** [comment]
I'd like to supply a patch, but I'm currently not sure what would be the right approach for this. Here are some options I see:

* make `notNull(Object[], String)` also check that the array is not null
* introduce `notEmpty(Object[], String)` with the same behavior as `notEmpty(Collection, String)`
* make `notEmpty(Collection, String)` also check that the collection does not contain null values
* introduce `notNull(Collection, String)` with the same behavior as `notNull(Object[], String)`
* introduce new methods for example `noneNull(Object[], String)`, `noneNull(Collection, String)` which check for not null, not empty and no null references contained

WDYT?

**Comment by sbrannen** [comment]
In order to (hopefully) avoid allowing myself to be biased by your proposals, I have analyzed the status quo and come up with the following proposals.

1. Rename `notNull(Object[], String)` to `notEmpty(Object[], String)`
1. Assert that the array passed to `notEmpty(Object[], String)` is not _empty_ (i.e., has length > 0)
1. Assert that all elements of the collection passed to `notEmpty(T, String)` are non-null

**Comment by sbrannen** [comment]
@britter, I edited your comment and converted the bullet points to numbered bullet points to make it easier to reason about the options.

**Comment by sbrannen** [comment]
My proposal effectively covers points 1, 2, and 3 from @britter.

Regarding points 4 and 5 from @britter...

I think we can do without the `notNull()` variants specific to arrays and collections since we already have general purpose `notNull(Object)` variants.

The `noneNull()` variants would be superseded by the now all encompassing functionality of `notEmpty(Object[], String)` and `notEmpty(Collection, String)`.

**Comment by sbrannen** [comment]
Hmmmmm.... I changed my mind: it's not as straightforward as it appears.

We have use cases where an _empty_ array is acceptable _but_ `null` values in the array are not acceptable.

So we can't simply get rid of `notNull(Object[], String)` like I originally thought/hoped.

**Comment by sbrannen** [comment]
@britter, I'm confused by your point # 1.

`notNull(Object[], String)` _already_ checks that the array is not `null`.

**Comment by sbrannen** [comment]
One thing to keep in mind is that we (i.e., the JUnit Team) do not want to introduce utilities that are never used, simply because, _as a general rule_, we don't support _dead code_.

In other words, although I see the merit of most of your proposals, @britter, I'm not sure we need all of those variants.

Thus, I will need to spend more time on this subject in order to properly assess our actual needs.

**Comment by britter** [comment]
> @britter, I'm confused by your point # 1.
Sorry, my mistake. I meant _not empty_. I've updated my comment.

**Comment by britter** [comment]
> One thing to keep in mind is that we (i.e., the JUnit Team) do not want to introduce utilities that are never used, simply because, as a general rule, we don't support dead code.

Full ACK

> In other words, although I see the merit of most of your proposals, @britter, I'm not sure we need all of those variants.

Maybe I put things the wrong way. My intend was to give some proposals to choose from.

> Thus, I will need to spend more time on this subject in order to properly assess our actual needs

Okay! Coming from the client code side (= the code using Preconditions) is probably a better idea. For example we have the strange behavior in TagsFilter where every empty/null combination throws an exception, expect from passing lists with null elements. I think a first step to move this issue forward is to add a check for null elements to `notEmpty(Collection)`.

**Comment by sbrannen** [comment]
> I think a first step to move this issue forward is to add a check for null elements to `notEmpty(Collection)`.

Yep, you're probably right in that regard.

**Comment by marcphilipp** [comment]
Now that #507 has been merged, what's the next step to move this forward?

**Comment by britter** [comment]
@marcphilipp I was thinking about providing the following methods to make the API sound:

* `notNull(Object[])`, `notNull(Collection)` -> Reference not null, contains no null references
* `notEmpty(Object[])`, `notEmpty(Cpllection)` -> Reference not null, contains no null references, is not empty

This way both use cases described by @sbrannen can be covered.

**Comment by bechte** [comment]
To me it looks like we have both cases:

- cases where we do not accept `null`, but accept `null` values within the array/collection
- cases where we do not accept `null` for the instance and values with the array/collection

Having a closer look at our implementation of `notNull(Collection)`, it clearly does not reflect what is stated in the JavaDoc: _"Assert that the supplied {@link Collection} is not {@code null} or empty."_

Having a closer look at our implementation of `notNull(Object[])`, it states something different in the JavaDoc: _"Assert that the supplied array and none of the objects it contains are `null`"_. In this case, we are assuming, that all our arrays are filled from 0 to n (n = size of array). We should not assume that the size of an array always matches the number of elements.

Having said all that, we might want to come up with three methods, that communicate their intend a little bit better:

1. `notNull(Object[])`, `notNull(Collection)` -> Reference not null, might contain null references
2. `notEmpty(Object[])`, `notEmpty(Collection)` -> Reference not null, contains at least one item that is not `null` (as for arrays, we do not know which position was filled with an value, see above).
3. `doesNotContainNull(Object[])`, doesNotContainNull(Collection)` -> Reference not null, contains no `null`, might be empty.

The names match the assertion names from AssertJ and they follow the same rule. See:
- org.assertj.core.internal.Iterables#assertNotEmpty
- org.assertj.core.internal.Iterables#assertDoesNotContainNull

This might be a good hint for us. What do you think?

**Comment by sbrannen** [comment]
I actually already decided (in my own brain over the past days) that we do in fact have 3 distinct use cases. So based on that principle, I agree with @bechte; however, I see a few things differently.

1. We don't need `notNull(Object[])` or `notNull(Collection)` since those are both covered by `notNull(Object)`.
1. The `notEmpty(Object[])` and `notEmpty(Collection)` variants should _only_ ensure that the array or collection is not null and contains at least one element; in other words, "not empty" should not assert anything about the actual elements.
1. I would call that method `containsNoNullElements(...)`. In addition, these variants should _only_ check that the array or collection does not contain null elements regardless of whether the array or collection is null itself or empty.

I think the above makes the APIs much more intuitive.

The only drawback is that there is not an "all in one" precondition for not null, not empty, and contains no null elements. But... I think that's a good thing. A developer can always check multiple preconditions to suit the use case at hand, and we in fact already do that in several places within the code base.

Thoughts?

**Comment by sbrannen** [comment]
Since I'm sure somebody will ask...

To address the original issue regarding discovery filters, with my above proposal the corresponding factory methods would have precondition checks similar to the following.

```java
Preconditions.notEmpty(tags, () -> "...");
Preconditions.containsNoNullElements(tags, () -> "...");
```

**Comment by britter** [comment]
> The only drawback is that there is not an "all in one" precondition for not null, not empty, and contains no null elements.

How about `isFilled(Object[]/Collection)` ? This is what I need for #493.

**Comment by jbduncan** [comment]
I personally agree with @sbrannen's proposed method names and behaviour.

It sounds pretty similar to how Guava does things regarding their own public Preconditions class and other classes like com.google.common.collect.CollectPreconditions, in that their method names are intuitive, clearly state what they're checking for, and don't try to do anything beyond what's expected.

**Comment by jbduncan** [comment]
I also think the "drawback" of not having an "all in one" precondition for not null, not empty and contains no null elements is not really be drawback at all. But then again I'm not in need of such a method myself, so my opinion here is probably not so meaningful. :P

**Comment by britter** [comment]
> [...] which may potentially catch users unaware if they don't read the javadoc.

As far as I understand the potential users are only the JUnit maintainers, since Preconditions should not be part of JUnits external API.

**Comment by sbrannen** [comment]
It affects not just the maintainers (i.e., core committers) but also contributors like yourself. 😉

**Comment by sbrannen** [comment]
@britter 

> How about isFilled(Object[]/Collection) ? This is what I need for #493.

That's equivalent to `notEmpty` and `containsNoNullElements` -- right?

**Comment by britter** [comment]
It's the "all in one" precondition (not null, not empty, and contains no null elements).

**Comment by sbrannen** [comment]
> It's the "all in one" precondition (not null, not empty, and contains no null elements).

Right!

... which is logically equivalent to `notEmpty` and `containsNoNullElements`, because _not empty_ implies _not null_. 😉 

So let me rephrase my question:

Could you live with using `notEmpty` and `containsNoNullElements` to achieve that goal?

**Comment by britter** [comment]
> Could you live with using notEmpty and containsNoNullElements to achieve that goal?

If the LauncherDiscoveryRequestBuilder is the only place where a combination of `notEmpty` and `containsNoNullElements` is needed, then I think it does not make sense to introduce a new method just for this use case. If we have more places where we need this kind of check, I think creating a specific method is justified to reduce duplicate code.

**Comment by sbrannen** [comment]
Sure.

Let's see how many places we actually need the "all in one" behavior.

**Comment by britter** [comment]
@sbrannen I had some time to look at this again. Your proposal was to drop `notNull(Object[], String)` and to introduce `notEmpty` and `containsNoNullElements`variants for both, arrays and collections.

The problem with removing `notNull(Object[], String)` is, that the behavior currently implemented can't be replaced by `notEmpty` and/or `containsNoNullElements`:

* `notNull(Object[], String)` -> array not null, contains no null elements
* `notEmpty` -> array not null, array.length > 0
* `containsNoNullElements` -> each element in array != null

We could replace every call to `notEmpty(Object[], String)` by calls to notEmpty and `containsNoNullElements`. But this would change the behavior by adding a check for length > 0. I can put a PR together with this change, but this may not be what we want.

Furthermore the already existing `notEmpty(Collection)` does check for not null, not empty and no null elements. If we make it consistent we will have to remove the no null elements check and add a `containsNoNullElement` call everywhere, `notEmpty(Collection)` is currently used.

this is getting complicated... :)

**Comment by bechte** [comment]
@britter 

One could replace `notNull(Object[], String)` by calling `notNull(Object, String)` and `containsNoNullElements(Object[], String)`. This would fit the need.

For the replacement of the current `notEmpty(Collection)`, one could do a little refactoring trick:

- rename the current method to `notEmptyOld(Collection)`
- introduce the new `notEmpty(Collection)`
- rewrite `notEmptyOld(Collection)` to call `notEmpty(Collection)` and `containsNoNullElements(Collection)` and _inline_ it everywhere

This will do the code changes automatically throughout the code base.

**Comment by jbduncan** [comment]
And just to make things even _more_ complicated (because I'm mean like that 😉), we may want to consider replacing the parameters of the `Preconditions.*(Collection)` methods with `Iterable` to expand their potential usefulness.

But considering that `Preconditions` is internal, this is may be premature until we know that we actually need to work with pure iterables rather than just collections.

**Comment by marcphilipp** [comment]
Team decision:

- Implement @sbrannen's proposal (see https://github.com/junit-team/junit5/issues/496#issuecomment-246975855)

**Comment by sbrannen** [comment]
This has been resolved in the three most recent commits to `master`.
